packages/studio-base/src/players/messageMemoryEstimation.test.ts

```diff
@@ -0,0 +1,98 @@
+    ).toThrow("Type 'UnknownType' not found in definitions");
+  });
+
+  it("handles complex types with arrays of unknown primitive type", () => {
+    const datatypes = new Map([
+      [
+        "ComplexType",
+        {
+          definitions: [{ type: "unknownType", name: "field1", isArray: true, arrayLength: 2 }],
+        },
+      ],
+    ]);
+
+    expect(() => estimateMessageObjectSize(datatypes, "ComplexType", new Map())).toThrow(
+      "Unknown primitive type unknownType",
+    );
+  });
+
+  it("size for a complex schema is calculated correctly", () => {
+    const datatypes = new Map([
+      [
+        "ComplexType",
+        {
+          definitions: [
+            { type: "int32", name: "field1" },
+            { type: "string", name: "field2" },
+          ],
+        },
+      ],
+    ]);
+
+    const sizeInBytes = estimateMessageObjectSize(datatypes, "ComplexType", new Map());
+    expect(sizeInBytes).toBeGreaterThan(0);
+  });
+
+  it("handles complex types with arrays", () => {
+    const datatypes = new Map([
+      [
+        "ComplexType",
+        {
+          definitions: [
+            { type: "int32", name: "field1", isArray: true, arrayLength: 3 },
+            { type: "string", name: "field2" },
+          ],
+        },
+      ],
+    ]);
+
+    const sizeInBytes = estimateMessageObjectSize(datatypes, "ComplexType", new Map());
+    expect(sizeInBytes).toBeGreaterThan(0);
+  });
+
+  it.each([
+    { numFloats: 2, numInts: 2, measuredSize: 52, tolerancePercent: 5 },
+    { numFloats: 10, numInts: 10, measuredSize: 212, tolerancePercent: 5 },
+    { numFloats: 10, numInts: 10, measuredSize: 212, tolerancePercent: 5 },
+    { numFloats: 50, numInts: 50, measuredSize: 1012, tolerancePercent: 5 },
+    { numFloats: 100, numInts: 100, measuredSize: 2012, tolerancePercent: 5 },
+    { numFloats: 200, numInts: 200, measuredSize: 4028, tolerancePercent: 5 },
+    { numFloats: 400, numInts: 400, measuredSize: 8024, tolerancePercent: 5 },
+    { numFloats: 550, numInts: 550, measuredSize: 31220, tolerancePercent: 5 },
+    { numFloats: 1000, numInts: 1000, measuredSize: 61196, tolerancePercent: 5 },
+    { numFloats: 2000, numInts: 2000, measuredSize: 122348, tolerancePercent: 5 },
+  ])(
+    "matches the size of objects with int + double fields measured with chrome devtools",
+    ({ numFloats, numInts, measuredSize, tolerancePercent }) => {
+      const datatypes = new Map([
+        [
+          "SomeType",
+          {
+            definitions: [
+              ...new Array(numInts).fill({ type: "int16", name: "field" }),
+              ...new Array(numFloats).fill({ type: "float64", name: "field" }),
+            ],
+          },
+        ],
+      ]);
+      const sizeInBytes = estimateMessageObjectSize(datatypes, "SomeType", new Map());
+      expect(sizeInBytes).toBeGreaterThanOrEqual(measuredSize * ((100 - tolerancePercent) / 100));
```

Comment:
can probably reduce these checks to one with:
```suggestion
      expect(Math.abs(sizeInBytes-measuredSize).toBeLessThanOrEqual(measuredSize * .01 * tolerancePercent);
```